### PR TITLE
Fit bonus

### PR DESCRIPTION
**Pull request by Jebzou**

### CODE REVIEW

File: ElectronicObserverTypes/Serialization/FitBonus/FitBonusData.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        /// Bonuses to apply
+        /// Applied x times, x being the number of equipment matching the conditions of the bonus fit 
+        /// If NumberOfEquipmentsRequiredAfterOtherFilters or EquipmentRequired or EquipmentTypesRequired, bonus is applied only once
+        /// </summary>
+        [JsonPropertyName("bonus")] public FitBonusValue Bonuses { get; set; } = null;
+
+        /// <summary>
+        /// Bonuses to apply if ship had a radar with LOS >= 5
+        /// </summary>
+        [JsonPropertyName("bonusSR")] public FitBonusValue BonusesIfLOSRadar { get; set; } = null;
```
Comment (myangelkamikaze) [inline ElectronicObserverTypes/Serialization/FitBonus/FitBonusData.cs]: They're called "surface radar" in our code base currently. That SR at the end is probably supposed to mean that, too. So I think `BonusesIfSurfaceRadar` would be a better name here.

File: ElectronicObserverTypes/Serialization/FitBonus/FitBonusPerEquipment.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace ElectronicObserverTypes.Serialization.FitBonus
+{
+    public class FitBonusPerEquipment
+    {
+        [JsonPropertyName("types")] public List<int>? EquipmentTypes { get; set; }
+
+        [JsonPropertyName("ids")] public List<int>? EquipmentIds { get; set; }
+
+        [JsonPropertyName("bonuses")] public List<FitBonusData>? Bonuses { get; set; }
```
Comment (myangelkamikaze) [inline ElectronicObserverTypes/Serialization/FitBonus/FitBonusPerEquipment.cs:12]: Should this be nullable? I'd expect this to always have a value.

File: ElectronicObserverTypes/Serialization/FitBonus/FitBonusValue.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+﻿using System.Text.Json.Serialization;
+
+namespace ElectronicObserverTypes.Serialization.FitBonus
+{
+    public class FitBonusValue
+    {
+
+        [JsonPropertyName("houg")] public int? Firepower { get; set; }
```
Comment (myangelkamikaze) [inline ElectronicObserverTypes/Serialization/FitBonus/FitBonusValue.cs:8]: Bonus values shouldn't be nullable, 0 works for no bonus.

File: ElectronicObserverTypes/Serialization/FitBonus/FitBonusData.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+        /// <summary>
+        /// Bonuses to apply if ship had a radar with AA >= 2
+        /// </summary>
+        [JsonPropertyName("bonusAR")] public FitBonusValue BonusesIfAirRadar { get; set; } = null;
```
Comment (myangelkamikaze) [inline ElectronicObserverTypes/Serialization/FitBonus/FitBonusData.cs]: This should be declared as nullable if null is a valid value. If this data is always available, make the default value either `new()` or `null!`. Same for every other property.

File: ElectronicObserver/Utility/Data/EquipmentFitBonus.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+	/// <summary>
+	/// Keep in mind that accuracy bonus is included
+	/// </summary>
+	/// <param name="ship"></param>
+	/// <returns></returns>
+	public static List<FitBonusValue> GetFitBonuses(this IShipData ship, IList<FitBonusPerEquipment> bonusList)
+	{
+		IList<IEquipmentData> equipments = ship.AllSlotInstance.Where(equipment => equipment != null).ToList()!;
+		if (ship.ExpansionSlotInstance != null) equipments.Add(ship.ExpansionSlotInstance);
```
Comment (myangelkamikaze) [inline ElectronicObserver/Utility/Data/EquipmentFitBonus.cs:27]: `AllSlotInstance` which just grabs the actual equipment from the ids in [AllSlot](https://github.com/gre4bee/ElectronicObserver/blob/main/ElectronicObserver/Data/ShipData.cs#L126), already contains the expansion slot.

File: ElectronicObserver/Utility/Data/EquipmentFitBonus.cs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	/// <param name="equipments"></param>
+	/// <param name="shipMaster"></param>
+	/// <param name="bonusList"></param>
+	/// <returns></returns>
+	public static List<FitBonusValue> GetFitBonuses(IList<IEquipmentData> equipments, IShipDataMaster shipMaster, IList<FitBonusPerEquipment> bonusList)
+	{
+		bool hasSurfaceRadar = equipments
+			.Any(equipment => equipment.MasterEquipment.IsSurfaceRadar);
+
+		bool hasAirRadar = equipments
```
Comment (myangelkamikaze) [inline ElectronicObserver/Utility/Data/EquipmentFitBonus.cs:44]: We have an extension for [HasAirRadar](https://github.com/gre4bee/ElectronicObserver/blob/main/ElectronicObserverTypes/Extensions/ShipDataExtensions.cs#L83).

File: ElectronicObserverTypes/Extensions/ShipDataExtensions.cs [hunk 7]
```diff
@@ -80,6 +80,8 @@ EquipmentTypes.MainGunLarge or
 	public static bool HasSurfaceRadar(this IShipData ship) => ship.AllSlotInstance
 		.Any(e => e?.MasterEquipment.IsSurfaceRadar is true);
 
+	public static bool HasAirRadar(this IShipData ship) => ship.HasAirRadar(1);
```
Comment (myangelkamikaze) [inline ElectronicObserverTypes/Extensions/ShipDataExtensions.cs:83]: It's probably better to just change the signature of HasAirRadar to:
public static bool HasAirRadar(this IShipData ship, int count = 1)

File: ElectronicObserverTypes/Serialization/FitBonus/FitBonusData.cs (hunk 4 above)
Comment (Jebzou) [reply ElectronicObserverTypes/Serialization/FitBonus/FitBonusData.cs]: Looking at the json every properties of that object can be actually null, even the bonuses

File: ElectronicObserverTypes/Serialization/FitBonus/FitBonusData.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
+﻿using System.Text.Json.Serialization;
+using System.Collections.Generic;
+
+namespace ElectronicObserverTypes.Serialization.FitBonus
```
Comment (myangelkamikaze) [inline ElectronicObserverTypes/Serialization/FitBonus/FitBonusData.cs:4]: Use file scoped namespaces. I think if you just add a `;` at the end of this line, Visual Studio should do the rest.

File: ElectronicObserverTypes/Serialization/FitBonus/FitBonusPerEquipment.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
+﻿using System.Collections.Generic;
+using System.Text.Json.Serialization;
+
+namespace ElectronicObserverTypes.Serialization.FitBonus
```
Comment (myangelkamikaze) [inline ElectronicObserverTypes/Serialization/FitBonus/FitBonusPerEquipment.cs:4]: File scoped namespace, and I think this file is mixing tabs and spaces. Make sure there are tabs everywhere.

**Pull request merged**